### ISSUE TITLE
Recursive PartialEq implementation not detected

### ISSUE THREAD

**Issue by 5225225** [rust-lang/rust#99220]
Given the following code:  https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ed9fb2f62239c9485d43d48e3e0d419b

```rust
struct Foo(u32);

impl std::cmp::PartialEq for Foo {
    fn eq(&self, rhs: &Self) -> bool {
        self == rhs
    }
}
```

The current output is:

```
```
(no warnings)

Ideally the output should look like:

```
warning: function cannot return without recursing
 --> src/lib.rs:4:5
  |
4 |     fn eq(&self, rhs: &Self) -> bool {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
5 |         self == rhs
  |         ----------- recursive call site
  |
  = note: `#[warn(unconditional_recursion)]` on by default
  = help: a `loop` may express intention better if this is on purpose
```

Note how we *do* warn about this if you do `self.eq(rhs)`. So I guess it's a matter of treating `==` as `.eq` for the purposes of recursion checking? Not sure.

Happens on nightly/beta/stable.

**Comment by tmiasko** [comment]
See also https://github.com/rust-lang/rust/issues/57299#issuecomment-451111538 and #57965.

**Comment by 5225225** [comment]
Ah, couldn't find those.

That being said, I think a special case for this body in particular might be worthwhile for as long as we don't have a more general solution, since I've seen people do it more than a few times.
